pkg/vmprovider/providers/vsphere/vmprovider.go

```diff
@@ -60,14 +76,36 @@ type vSphereVMProvider struct {
 func NewVSphereVMProviderFromClient(
 	client ctrlruntime.Client,
 	recorder record.Recorder) vmprovider.VirtualMachineProviderInterface {
+	ovfCache, ovfLockPool := InitOvfCacheAndLockPool(
+		ovfCacheItemExpiration, ovfCacheExpirationCheckInterval, ovfCacheMaxItem)
 
 	return &vSphereVMProvider{
 		k8sClient:         client,
 		eventRecorder:     recorder,
 		globalExtraConfig: getExtraConfig(),
+		ovfCache:          ovfCache,
+		ovfCacheLockPool:  ovfLockPool,
 	}
 }
 
+// InitOvfCacheAndLockPool initializes the ovf cache and lock pool that are used
+// to cache the ovf envelope and lock the ovf envelope when it is being downloaded.
+func InitOvfCacheAndLockPool(expireAfter, checkExpireInterval time.Duration, maxItems int) (
+	*util.Cache[VersionedOVFEnvelope], *util.LockPool[string, *sync.RWMutex]) {
+	ovfCache := util.NewCache[VersionedOVFEnvelope](expireAfter, checkExpireInterval, maxItems)
+	ovfLockPool := &util.LockPool[string, *sync.RWMutex]{}
+
+	// Clean up the lock pool when the ovf cache item expires.
+	expiredChan := ovfCache.ExpiredChan()
+	go func() {
+		for k := range expiredChan {
```

Comment:
Isn't this racey: before the message is handled here, another call comes into getOvfEnvelope() and then an in-use lock is removed from this cache?

Reply:
Yeah, but that shouldn't block us because the in-use lock can still be safely executed after removal, although it won't prevent another download of the same item. I considered locking it during deletion, but it wouldn't prevent the lock from being used after removal. Adding another layer of protection for the lock pool isn't ideal either, as there could be at most 2 duplicate downloads in the worst-case scenario. And this only occurs in rare cases when the lock is removed while in use.

Reply:
On second thoughts, maybe locking and unlocking during deletion could still be useful. If getOvfEnvelope() can lock it, the lock won't be deleted until the item is downloaded and put into the cache, avoiding duplicate downloads. However, if delete() locks it first, there's still a small chance of duplicate downloads. I'll add comments in the updated code to document our discussion here.

Reply:
Yes, it will break the guarantee of no duplicate OVF retrieves we're trying to achieve in the first place by adding this outer locking pool.

Ultimately I think we're trying to make something more complicated that what's needed: the outer lock should be pushed into the CacheItem[T], and the Cache[T] has a Load/Fill function that populates the T, so we have something like https://guava.dev/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html, and there's no need for the eviction channel to try to maintain consistency of the other cache.

Reply:
I believe that's what I had in my original content library specific cache, where the item has its own lock... However, I understand this util cache doesn't have item-level locks as it may not be necessary for common use cases, where the cache is populated or retrieved quickly. It may not completely eliminate duplicate OVF retrieves, but it only occurs in rare edge cases where an OVF expires right after starting to download, and only two duplicates could happen even with multiple namespaces associated. If you feel strongly about this, I'm happy to discuss it further offline and make a follow-up change to address it.